pkg/parser/modifier/matrix/matrix.go

```diff
@@ -14,84 +14,60 @@ var ErrMatrixNeedsCollection = errors.New("matrix should contain a collection")
 // something other than maps (e.g. lists or scalars) as it's items.
 var ErrMatrixNeedsListOfMaps = errors.New("matrix with a list can only contain maps as it's items")
 
-// errExpansionCommenced is returned when the matrix modifier gets expanded
-// and we need to do another pass to see if there's more.
-var errExpansionCommenced = errors.New("single expansion commenced")
+// ErrMatrixIsMisplaced is returned when the matrix modifier is attached
+// to a task type other than task or docker_builder.
+var ErrMatrixIsMisplaced = errors.New("matrix can be defined only under a task or docker_builder")
 
 // errNoExpansionDone is returned when a single pass yields no matrix expansions.
 var errNoExpansionDone = errors.New("no matrix expansion was done")
 
-// Recursively processes each "outer" map key of the loaded YAML document
-// in an attempt to produce multiple keys as a result of matrix expansion.
-func singlePass(inputTree yaml.MapSlice) (yaml.MapSlice, error) {
-	var outputTree yaml.MapSlice
-	var atLeastOneExpansion bool
+func singlePass(inputTree *node.Node) error {
+	// Find a matrix node to expand
+	matrixNode := inputTree.DeepFindChild("matrix")
+	if matrixNode == nil {
+		return errNoExpansionDone
+	}
 
-	if len(inputTree) == 0 {
-		return nil, errNoExpansionDone
+	// Ensure this matrix node is attached to either a task or a docker_builder
```

Comment:
I wonder if it should work with `pipe` 🤔

Reply:
Should it?